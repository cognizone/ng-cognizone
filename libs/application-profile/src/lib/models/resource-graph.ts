import { TypedResourceGraph } from '@cognizone/model-utils';

/**
 * @deprecated use TypedResourceGraph from @cognizone/model-utils
 */
export type ResourceGraphRaw = TypedResourceGraph;
