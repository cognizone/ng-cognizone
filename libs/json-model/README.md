# json-model

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test json-model` to execute the unit tests.
