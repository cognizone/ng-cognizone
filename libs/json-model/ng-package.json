{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/json-model",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
