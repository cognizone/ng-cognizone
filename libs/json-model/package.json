{
  "name": "@cognizone/json-model",
  "version": "3.7.1",
  "peerDependencies": {
    "@cognizone/lod": ">=3.0.0",
    "@cognizone/model-utils": ">=3.0.0",
    "@cognizone/ng-core": ">=3.0.0"
  },
  "dependencies": {
    "tslib": "^2.2.0"
  },
  "license": "Apache-2.0",
  "homepage": "https://github.com/cognizone/ng-cognizone",
  "contributors": [
    {
      "name": "William Chelman",
      "email": "william.chelman@cogni.zone"
    },
    {
      "name": "Farah El Agha",
      "email": "farah.agha111@gmail.com"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cognizone/ng-cognizone.git",
    "directory": "libs/json-model"
  },
  "private": false,
  "publishConfig": {
    "access": "public"
  }
}
