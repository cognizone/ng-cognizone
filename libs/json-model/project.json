{
  "name": "json-model",
  "projectType": "library",
  "sourceRoot": "libs/json-model/src",
  "prefix": "cz",
  "targets": {
    "build": {
      "executor": "@nrwl/angular:package",
      "outputs": ["{workspaceRoot}/dist/libs/json-model"],
      "options": {
        "project": "libs/json-model/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/json-model/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/json-model/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/json-model"],
      "options": {
        "jestConfig": "libs/json-model/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": ["libs/json-model/src/**/*.ts", "libs/json-model/src/**/*.html"]
      },
      "outputs": ["{options.outputFile}"]
    }
  },
  "tags": []
}
