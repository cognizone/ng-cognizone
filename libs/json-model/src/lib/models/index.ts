export * from './json-model';
export * from './resource';
