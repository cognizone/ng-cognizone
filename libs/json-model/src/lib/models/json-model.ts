/* eslint-disable @typescript-eslint/naming-convention */
import { Many, TypedResourceContext } from '@cognizone/model-utils';

/**
 * @experimental
 */
export type Uri<T extends JsonModel> = string & { '@@meta_type_placeholder@@'?: T };

export interface JsonModel {
  '@id': string;
  '@type': JsonModelType;
  '@context'?: TypedResourceContext;
  '@facets'?: {};
}

export type JsonModelType<T extends Many<string> = string> = T extends string ? Many<string> | T | [T, ...string[]] : Many<string>;

export function isJsonModel(o: unknown): o is JsonModel {
  return typeof o === 'object' && o != null && '@id' in o && '@type' in o;
}

export interface JsonModels<T extends JsonModel = JsonModel> {
  [uri: string]: T;
}

export type JsonModelFlat<T = JsonModel> = {
  [key in keyof T]: T[key] extends JsonModel
    ? string
    : T[key] extends JsonModel | undefined
    ? string | undefined
    : T[key] extends JsonModel[]
    ? string[]
    : T[key] extends JsonModel[] | undefined
    ? string[] | undefined
    : T[key];
};

export interface JsonModelFlatGraph<T extends JsonModelFlat = JsonModelFlat> {
  rootUri: string;
  context?: TypedResourceContext;
  facets?: {};
  models: {
    [uri: string]: T;
  };
}

export interface JsonModelValue<T> {
  '@value': T;
  '@type': string;
}

export function isJsonModelValue<T = unknown>(o: unknown): o is JsonModelValue<T> {
  return typeof o === 'object' && o != null && '@type' in o && '@value' in o;
}
