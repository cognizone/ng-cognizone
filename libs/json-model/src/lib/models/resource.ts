import { Many, TypedResourceContext } from '@cognizone/model-utils';

/**
 * @description intermediary model used internally to convert JsonModel to
 * TypedResourceGraph (and the other way around)
 */
export interface ResourceGraph {
  data: Resource;
  included: Resource[];
  facets?: {};
  context?: TypedResourceContext;
}

/**
 * @description intermediary model used internally to convert JsonModel to
 * TypedResourceGraph (and the other way around)
 */
export interface Resource<T extends object = {}> {
  uri: string;
  type: Many<string>;
  references: { [referenceKey: string]: Many<string> }; // value is one or multiple uris
  attributes: { [P in keyof T]: Many<ResourceAttribute<T[P]>> };
}

/**
 * @description intermediary model used internally to convert JsonModel to
 * TypedResourceGraph (and the other way around)
 */
export interface ResourceAttribute<T> {
  value: T;
  dataType: string;
}
