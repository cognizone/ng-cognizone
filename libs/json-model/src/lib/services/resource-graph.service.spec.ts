/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import { TypedResourceGraph } from '@cognizone/model-utils';
import { LoggerModule } from '@cognizone/ng-core';
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import { ApplicationProfile } from '@cognizone/application-profile';
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import { NgApplicationProfileModule } from '@cognizone/ng-application-profile';
import { JsonModel } from '../models/json-model';

import { JsonModelModule } from '../json-model.module';
import { ResourceGraphService } from './resource-graph.service';

const directiveAp: ApplicationProfile = require('../../test/directive-ap.json');
const treatyAp: ApplicationProfile = require('../../test/treaty-ap.json');
const draftAp: ApplicationProfile = require('../../test/lux-legipro-ap.json');
const treatyProcess1: JsonModel = require('../../test/data/treaty-json-model-1.json');
const rawResourceGraph1: TypedResourceGraph = require('../../test/data/casemates-concept-1.json');
const rawResourceGraph2: TypedResourceGraph = require('../../test/data/directives-eudossier-2.json');
const rawResourceGraph3: TypedResourceGraph = require('../../test/data/directives-eudossier-1.json');
const rawResourceGraph4: TypedResourceGraph = require('../../test/data/legiswiss-act-1.json');
const rawDraft1: TypedResourceGraph = require('../../test/data/draft-1.json');
const rawDraft2: TypedResourceGraph = require('../../test/data/draft-2.json');
const rawDraft3: TypedResourceGraph = require('../../test/data/draft-3.json');

describe('ResourceGraphService', () => {
  const createService = createServiceFactory({
    service: ResourceGraphService,
    imports: [JsonModelModule.forRoot(), NgApplicationProfileModule.forRoot(), LoggerModule.forRoot('TEST')],
  });

  let spectator: SpectatorService<ResourceGraphService>;

  beforeEach(() => (spectator = createService()));

  it('should be injectable', () => {
    expect(spectator.service).not.toBeUndefined();
  });

  it('should transform ResourceGraph 1 to JsonModel', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph1);
    expect(json).toMatchSnapshot();
  });

  it('should transform ResourceGraph 2 to JsonModel', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph2);
    expect(json).toMatchSnapshot();
  });

  it('should deserialize then serialize ResourceGraph 2', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph2);
    const newRaw = spectator.service.jsonModelToResourceGraphRaw(json, directiveAp);
    expect(newRaw).toEqual(rawResourceGraph2);
  });

  it('should transform ResourceGraph 2 and flatten it with ap', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph2, directiveAp) as JsonModel & {
      eventRiskLevel: string;
    };
    expect(json.eventRiskLevel).toBe('http://resource/my-domain/6');
  });

  it('should transform ResourceGraph 2 and not it flatten without ap', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph2) as JsonModel & {
      eventRiskLevel: string[];
    };
    expect(json.eventRiskLevel).toMatchObject(['http://resource/my-domain/6']);
  });

  it('should transform ResourceGraph 3 to JsonModel', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph3);
    expect(json).toMatchSnapshot();
  });

  it('should deserialize then serialize ResourceGraph 3', () => {
    const jsonLd = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph3);
    const newRaw = spectator.service.jsonModelToResourceGraphRaw(jsonLd, directiveAp);
    expect(newRaw).toEqual(rawResourceGraph3);
  });

  it('should transform ResourceGraph 4 to JsonModel', () => {
    const jsonLd = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph4);
    expect(jsonLd).toMatchSnapshot();
  });

  it('should transform ResourceGraph 5 to JsonModel (circular graph)', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph5);
    expect(json).toMatchSnapshot();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect((json as any).child.child.grandParent).toBe(json);
  });

  it('should deserialize then serialize ResourceGraph 5 (circular graph)', () => {
    const json = spectator.service.resourceGraphRawToJsonModel(rawResourceGraph5);
    const newRaw = spectator.service.jsonModelToResourceGraphRaw(json, ap5);
    expect(newRaw).toEqual(rawResourceGraph5);
  });

  it('should deserialize big json', () => {
    const json = require('../../test/legal-taxonomy.json').hits.hits[0]._source;
    const model = spectator.service.resourceGraphRawToJsonModel(json);
    // we just test that it succeeded
    expect(model).toBeTruthy();
  });

  it('should serialize treaty process 1', () => {
    const model = spectator.service.jsonModelToResourceGraphRaw(treatyProcess1, treatyAp);
    model?.included?.forEach(d => {
      expect(JSON.stringify(d) !== '');
    });
    expect(model).toMatchSnapshot();
  });

  it('should serialize treaty and remove unknown attribute', () => {
    jest.spyOn(console, 'warn').mockImplementation();
    const model = {
      '@id': 'https://fedlex.data.admin.ch/treatyProcess/9999/5414',
      '@type': 'TreatyProcess',
      treatyProcessHasResultingTreatyDocument: [],
      treatyCommentAboutValidity: [],
      treatySubject: [],
      treatyLimitDateForSignature: [],
      institutionInChargeOfTheTreatyLevel2: [],
      treatyComment: [],
      treatyHasRatificationRestriction: [],
      treatyNotificationPublishedInLabel: [],
      treatyDescription: [],
      treatyHasEntryIntoForceScope: [],
      treatyNotificationPublishedIn: [],
      committeeCreationBasedOn: [],
      treatyLimitDateForSignatureComment: [],
      treatyPartyCountry: [],
      titleTreaty: {},
      treatyPartyOrganisation: [],
      treatyAppliesEULegalResource: [
        {
          '@id': 'http://publications.europa.eu/resource/cellar/d67cf7eb-de13-4bcb-9959-0dc0107e102e',
          '@type': 'EULegalResource',
        },
      ],
      treatyRelatedLegalResource: [],
      confidentialTreaty: undefined,
      treatyInForceStatusForOurCountry: undefined,
      institutionInChargeOfTheTreaty: undefined,
      treatyConsolidatedBy: undefined,
      needForOfficialPublication: undefined,
      treatyLegalResourceForOtherParty: undefined,
      treatyOtherPublicationReference: undefined,
      anAttributeThatIsNowhereToBeFound: [4, 5, 6], // this should not be serialized
    };

    const raw = spectator.service.jsonModelToResourceGraphRaw(model, treatyAp);
    expect(raw).toMatchSnapshot();
    expect(console.warn).toHaveBeenCalled();
  });

  describe.only('or range rule on datatype', () => {
    it('should deserialize then serialize draft 1 (or on rapporteurChambreDeputesLabel, xsd:string)', () => {
      const json = spectator.service.resourceGraphRawToJsonModel(rawDraft1, draftAp);
      const newRaw = spectator.service.jsonModelToResourceGraphRaw(json, draftAp);
      checkRawEquality(newRaw, rawDraft1, ['internalId']);
    });

    it('should deserialize then serialize draft 2 (or on rapporteurChambreDeputesLabel, xsd:string)', () => {
      const json = spectator.service.resourceGraphRawToJsonModel(rawDraft2, draftAp);
      const newRaw = spectator.service.jsonModelToResourceGraphRaw(json, draftAp);
      checkRawEquality(newRaw, rawDraft2, ['internalId']);
    });

    it('should deserialize then serialize draft 3 (or on rapporteurChambreDeputesLabel, rdfs:Resource)', () => {
      const json = spectator.service.resourceGraphRawToJsonModel(rawDraft3, draftAp);
      const newRaw = spectator.service.jsonModelToResourceGraphRaw(json, draftAp);
      checkRawEquality(newRaw, rawDraft3, ['internalId']);
    });

    it('should deserialize then serialize draft 4 (or on rapporteurChambreDeputesLabel, rdfs:Resource)', () => {
      const json = spectator.service.resourceGraphRawToJsonModel(rawDraft3, draftAp);
      const newRaw = spectator.service.jsonModelToResourceGraphRaw(json, draftAp);
      checkRawEquality(newRaw, rawDraft3, ['internalId']);
    });
  });
});

// circular graph
const rawResourceGraph5: TypedResourceGraph = {
  data: {
    uri: 'uri1',
    type: 'root',
    references: {
      child: 'uri2',
    },
  },
  included: [
    {
      uri: 'uri2',
      type: 'root',
      references: {
        child: 'uri3',
      },
    },
    {
      uri: 'uri3',
      type: 'grandChild',
      references: {
        grandParent: 'uri1',
      },
    },
  ],
};

const ap5: ApplicationProfile = {
  uri: 'ap5',
  types: {
    root: {
      classIds: ['root'],
      rules: [{ name: 'rdfTypes', value: ['root'] }],
      attributes: {
        child: {
          attributeId: 'child',
          rules: [{ name: 'range', value: { name: 'classId', value: 'root' } }],
          uri: 'uri:child',
        },
      },
    },
    grandChild: {
      classIds: ['grandChild'],
      rules: [{ name: 'rdfTypes', value: ['grandChild'] }],
      attributes: {
        grandParent: {
          attributeId: 'grandParent',
          rules: [{ name: 'range', value: { name: 'classId', value: 'root' } }],
          uri: 'uri:grandParent',
        },
      },
    },
  },
};

function checkRawEquality(x: TypedResourceGraph | undefined, y: TypedResourceGraph | undefined, ignoredAttributes: string[] = []): void {
  const getSortedNodes = (a: TypedResourceGraph) =>
    [a.data, ...(a.included ?? [])]
      .sort((b, c) => b.uri.localeCompare(c.uri))
      .map(node => {
        const newNode = { ...node, rootType: undefined };
        if (newNode.attributes) {
          ignoredAttributes.forEach(field => delete newNode.attributes?.[field]);
        }
        return newNode;
      });

  const allX = x ? getSortedNodes(x) : [];
  const allY = y ? getSortedNodes(y) : [];

  expect(allX.length).toEqual(allY.length);
  expect(allX).toEqual(allY);
}
