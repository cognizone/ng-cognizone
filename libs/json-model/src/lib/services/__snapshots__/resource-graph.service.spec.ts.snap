// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceGraphService should serialize treaty and remove unknown attribute 1`] = `
Object {
  "data": Object {
    "attributes": Object {
      "treatyComment": Object {
        "xsd:string": Array [],
      },
      "treatyCommentAboutValidity": Object {
        "xsd:string": Array [],
      },
      "treatyLimitDateForSignature": Object {
        "xsd:date": Array [],
      },
      "treatyLimitDateForSignatureComment": Object {
        "xsd:string": Array [],
      },
      "treatyNotificationPublishedIn": Object {
        "rdfs:Resource": Array [],
      },
      "treatyNotificationPublishedInLabel": Object {
        "xsd:string": Array [],
      },
    },
    "references": Object {
      "treatyAppliesEULegalResource": Array [
        "http://publications.europa.eu/resource/cellar/d67cf7eb-de13-4bcb-9959-0dc0107e102e",
      ],
    },
    "type": "TreatyProcess",
    "uri": "https://fedlex.data.admin.ch/treatyProcess/9999/5414",
  },
  "included": Array [
    Object {
      "type": "EULegalResource",
      "uri": "http://publications.europa.eu/resource/cellar/d67cf7eb-de13-4bcb-9959-0dc0107e102e",
    },
  ],
}
`;

exports[`ResourceGraphService should serialize treaty process 1 1`] = `
Object {
  "context": Object {
    "base": "http://example.org/",
    "prefix": Object {
      "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
      "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
      "xsd": "http://www.w3.org/2001/XMLSchema#",
    },
  },
  "data": Object {
    "attributes": Object {
      "bilateral": Object {
        "xsd:boolean": true,
      },
      "confidentialTreaty": Object {
        "xsd:boolean": false,
      },
      "treatyComment": Object {
        "xsd:string": Array [],
      },
      "treatyCommentAboutValidity": Object {
        "xsd:string": Array [],
      },
      "treatyNotificationPublishedIn": Object {
        "rdfs:Resource": Array [],
      },
      "treatyNotificationPublishedInLabel": Object {
        "xsd:string": Array [],
      },
      "treatyProcessIdentificationInCms": Object {
        "xsd:string": Array [],
      },
    },
    "references": Object {
      "treatyProcessHasResultingTreatyDocument": Array [
        "http://resource/TreatyDocument/3-1591712645034",
        "http://resource/TreatyDocument/8-1591712649311",
      ],
      "treatyProcessHasTask": Array [
        "http://resource/TreatySignature/1-1591712631407",
        "http://resource/EntryIntoForce/2-1591712631410",
      ],
      "treatyType": "http://resource/my-domain/1",
    },
    "type": "TreatyProcess",
    "uri": "http://resource/TreatyProcess/0-1591712631312",
  },
  "facets": Object {
    "someTitle": "I love titles",
  },
  "included": Array [
    Object {
      "references": Object {
        "typeDocument": "https://fedlex.data.admin.ch/vocabulary/resource-type/31",
      },
      "type": "TreatyDocument",
      "uri": "http://resource/TreatyDocument/3-1591712645034",
    },
    Object {
      "type": "ResourceType",
      "uri": "https://fedlex.data.admin.ch/vocabulary/resource-type/31",
    },
    Object {
      "references": Object {
        "isRealizedBy": Array [
          "http://resource/Expression/7-1591712649311",
        ],
      },
      "type": "TreatyDocument",
      "uri": "http://resource/TreatyDocument/8-1591712649311",
    },
    Object {
      "attributes": Object {
        "title": Object {
          "xsd:string": "there-it-is.gif",
        },
      },
      "references": Object {
        "isEmbodiedBy": Array [
          "http://resource/Manifestation/6-1591712649311",
        ],
      },
      "type": "Expression",
      "uri": "http://resource/Expression/7-1591712649311",
    },
    Object {
      "attributes": Object {
        "fileName": Object {
          "xsd:string": "there-it-is.gif",
        },
        "isExemplifiedBy": Object {
          "rdfs:Resource": "http://resource/Resource/5-1591712649311",
        },
        "uploadBy": Object {
          "xsd:string": "John Doe",
        },
      },
      "references": Object {
        "uploadDate": "2020-06-09T14:24:09.311Z",
      },
      "type": "Manifestation",
      "uri": "http://resource/Manifestation/6-1591712649311",
    },
    Object {
      "attributes": Object {
        "languageLabel": Object {
          "xsd:string": Array [],
        },
        "treatyComment": Object {
          "xsd:string": Array [],
        },
      },
      "type": "TreatySignature",
      "uri": "http://resource/TreatySignature/1-1591712631407",
    },
    Object {
      "type": "EntryIntoForce",
      "uri": "http://resource/EntryIntoForce/2-1591712631410",
    },
    Object {
      "type": "TreatyType",
      "uri": "http://resource/my-domain/1",
    },
  ],
}
`;

exports[`ResourceGraphService should transform ResourceGraph 1 to JsonModel 1`] = `
Object {
  "@facets": Object {
    "some": "thing",
  },
  "@id": "http://resource/my-domain/1",
  "@type": "Concept",
  "created": Array [
    2022-02-01T15:55:23.473Z,
  ],
  "definition": Object {
    "en": Array [
      "en label5",
    ],
  },
  "inScheme": Array [
    Object {
      "@id": "http://resource/my-domain/2",
      "@type": "ConceptScheme",
    },
  ],
  "prefLabel": Object {
    "en": Array [
      "en label4",
    ],
  },
  "status": Array [
    Object {
      "@id": "http://resource/my-domain/3",
      "@type": "Concept",
      "authorityCode": Array [
        "string-9",
      ],
      "identifier": Array [
        "string-7",
      ],
      "inScheme": Array [
        Object {
          "@id": "http://resource/my-domain/6",
          "@type": "ConceptScheme",
        },
      ],
      "opCode": Array [
        "string-11",
      ],
      "prefLabel": Object {
        "en": Array [
          "en label8",
        ],
      },
      "startUse": Array [
        "string-10",
      ],
    },
  ],
  "topConceptOf": Array [
    Object {
      "@id": "http://resource/my-domain/2",
      "@type": "ConceptScheme",
    },
  ],
}
`;

exports[`ResourceGraphService should transform ResourceGraph 2 to JsonModel 1`] = `
Object {
  "@id": "http://resource/my-domain/1",
  "@type": Array [
    "EUDossier",
    "Event",
  ],
  "eventRiskLevel": Array [
    "http://resource/my-domain/6",
  ],
  "identifier": Array [
    "string-5",
  ],
  "initiatedByActPreparatory": Array [
    Object {
      "@id": "http://resource/my-domain/2",
      "@type": Array [
        "EUPreparatoryAct",
        "Work",
      ],
      "dateDocument": Array [
        2022-02-01T00:00:00.000Z,
      ],
      "preparatoryActCelexId": Array [
        "string-9",
      ],
      "preparatoryActReference": Array [
        "string-10",
        "string-11",
        "string-12",
        "string-13",
      ],
      "responsibilityOfEUInstitution": Array [
        Object {
          "@id": "http://resource/my-domain/8",
          "@type": Array [
            "Concept",
            "EULegalInstitution",
          ],
        },
      ],
      "title": Object {
        "en": Array [
          "en label14",
        ],
      },
    },
  ],
  "initiatesTransposition": Object {
    "@id": "http://resource/my-domain/3",
    "@type": Array [
      "Event",
      "Transposition",
    ],
    "hasTranspositionAction": Array [
      Object {
        "@id": "http://resource/my-domain/15",
        "@type": Array [
          "Event",
          "TranspositionAction",
        ],
        "created": 2022-02-01T16:04:01.159Z,
        "transpositionActionType": Object {
          "@id": "http://resource/my-domain/16",
          "@type": Array [
            "Concept",
            "TranspositionActionType",
          ],
        },
      },
    ],
    "transpositionRequestedCorrespondenceTable": false,
  },
  "isAboutConceptTypeDossier": Array [
    Object {
      "@id": "http://resource/my-domain/4",
      "@type": Array [
        "Concept",
        "EUDossierType",
      ],
    },
  ],
  "title": Object {
    "en": Array [
      "en label7",
    ],
  },
}
`;

exports[`ResourceGraphService should transform ResourceGraph 3 to JsonModel 1`] = `
Object {
  "@id": "http://resource/my-domain/1",
  "@type": Array [
    "EUDossier",
    "Event",
  ],
  "identifier": Array [
    "string-5",
  ],
  "initiatedByActPreparatory": Array [
    Object {
      "@id": "http://resource/my-domain/2",
      "@type": Array [
        "EUPreparatoryAct",
        "Work",
      ],
      "dateDocument": Array [
        2022-02-01T00:00:00.000Z,
      ],
      "preparatoryActCelexId": Array [
        "string-15",
      ],
      "preparatoryActReference": Array [
        "string-16",
        "string-17",
        "string-18",
      ],
      "preparatoryActSubject": Array [
        Object {
          "@id": "http://resource/my-domain/7",
          "@type": "Concept",
        },
        Object {
          "@id": "http://resource/my-domain/8",
          "@type": "Concept",
        },
        Object {
          "@id": "http://resource/my-domain/9",
          "@type": "Concept",
        },
        Object {
          "@id": "http://resource/my-domain/10",
          "@type": "Concept",
        },
        Object {
          "@id": "http://resource/my-domain/11",
          "@type": "Concept",
        },
        Object {
          "@id": "http://resource/my-domain/12",
          "@type": "Concept",
        },
        Object {
          "@id": "http://resource/my-domain/13",
          "@type": "Concept",
        },
      ],
      "responsibilityOfEUInstitution": Array [
        Object {
          "@id": "http://resource/my-domain/14",
          "@type": Array [
            "Concept",
            "EULegalInstitution",
          ],
        },
      ],
      "title": Object {
        "en": Array [
          "en label19",
        ],
      },
    },
  ],
  "initiatesTransposition": Object {
    "@id": "http://resource/my-domain/3",
    "@type": Array [
      "Event",
      "Transposition",
    ],
    "hasTranspositionAction": Array [
      Object {
        "@id": "http://resource/my-domain/20",
        "@type": Array [
          "Event",
          "TranspositionAction",
        ],
        "created": 2022-02-01T15:56:07.288Z,
        "transpositionActionType": Object {
          "@id": "http://resource/my-domain/21",
          "@type": Array [
            "Concept",
            "TranspositionActionType",
          ],
        },
      },
    ],
    "transpositionRequestedCorrespondenceTable": false,
  },
  "isAboutConceptTypeDossier": Array [
    Object {
      "@id": "http://resource/my-domain/4",
      "@type": Array [
        "Concept",
        "EUDossierType",
      ],
    },
  ],
  "title": Object {
    "en": Array [
      "en label6",
    ],
  },
}
`;

exports[`ResourceGraphService should transform ResourceGraph 4 to JsonModel 1`] = `
Object {
  "@id": "http://resource/my-domain/1",
  "@type": Array [
    "Act",
    "Work",
  ],
  "dateDocument": Array [
    2022-02-01T00:00:00.000Z,
  ],
  "dateEntryInForce": Array [
    2022-02-01T00:00:00.000Z,
  ],
  "isRealizedBy": Array [
    Object {
      "@id": "http://resource/my-domain/2",
      "@type": "Expression",
      "isEmbodiedBy": Array [
        Object {
          "@id": "http://resource/my-domain/5",
          "@type": "Manifestation",
          "isExemplifiedBy": Array [
            "http://resource/my-domain/11",
          ],
        },
        Object {
          "@id": "http://resource/my-domain/6",
          "@type": "Manifestation",
          "isExemplifiedBy": Array [
            "http://resource/my-domain/12",
          ],
        },
      ],
    },
    Object {
      "@id": "http://resource/my-domain/3",
      "@type": "Expression",
      "isEmbodiedBy": Array [
        Object {
          "@id": "http://resource/my-domain/7",
          "@type": "Manifestation",
          "isExemplifiedBy": Array [
            "http://resource/my-domain/13",
          ],
        },
        Object {
          "@id": "http://resource/my-domain/8",
          "@type": "Manifestation",
          "isExemplifiedBy": Array [
            "http://resource/my-domain/14",
          ],
        },
      ],
    },
    Object {
      "@id": "http://resource/my-domain/4",
      "@type": "Expression",
      "isEmbodiedBy": Array [
        Object {
          "@id": "http://resource/my-domain/9",
          "@type": "Manifestation",
          "isExemplifiedBy": Array [
            "http://resource/my-domain/15",
          ],
        },
        Object {
          "@id": "http://resource/my-domain/10",
          "@type": "Manifestation",
          "isExemplifiedBy": Array [
            "http://resource/my-domain/16",
          ],
        },
      ],
    },
  ],
  "publicationDate": Array [
    2022-02-01T00:00:00.000Z,
  ],
}
`;

exports[`ResourceGraphService should transform ResourceGraph 5 to JsonModel (circular graph) 1`] = `
Object {
  "@id": "uri1",
  "@type": "root",
  "child": Object {
    "@id": "uri2",
    "@type": "root",
    "child": Object {
      "@id": "uri3",
      "@type": "grandChild",
      "grandParent": [Circular],
    },
  },
}
`;
